sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -187,6 +187,9 @@ class Dataset[T] private[sql](
     }
   }
 
+  // Wrap analyzed logical plan with an analysis barrier so we won't traverse/resolve it again.
+  @transient private val planWithBarrier: LogicalPlan = AnalysisBarrier(logicalPlan)
```

Comment:
`CacheManager` uses `Dataset.logicalPlan` as key to look up identical plans already cached. If we always wrap `logicalPlan` with a barrier, we need to strip it when looking up caches.